### PR TITLE
wxWindowsPrintNativeData InitializeDevMode fix

### PR DESCRIPTION
**Pull request by sz5000**
Use a temporary WinPrinter if not specified via unction call in order to not fall back to the default printer code.

Do not use PRINTDLG if a printerName is specified because this returns the default printer which is not necessarily the printer we want. Therefore we have to ensure, that the code above is used and therefore we use a temporary printer if not specified.

### CODE REVIEW

File: src/msw/printdlg.cpp [hunk 1]
```diff
@@ -423,6 +423,12 @@ void wxWindowsPrintNativeData::InitializeDevMode(const wxString& printerName, Wi
     // this replaces the PrintDlg function which creates the DEVMODE filled only with data from default printer.
     if ( !m_devMode && !printerName.IsEmpty() )
     {
+        // ensure that we have a printer object here, otherwise we are unable to determine m_devMode
+        WinPrinter fallbackPrinter;
+        if (!printer) {
+            printer = &fallbackPrinter;
+        }
```
Comment (vadz) [inline src/msw/printdlg.cpp:430]: This should be reformatted to put braces on the correct lines or just remove them:
```suggestion
        if (!printer)
            printer = &fallbackPrinter;
```

File: src/msw/printdlg.cpp [hunk 2]
```diff
@@ -423,6 +423,12 @@ void wxWindowsPrintNativeData::InitializeDevMode(const wxString& printerName, Wi
...
+
         // Open printer
         if ( printer && printer->Open( printerName ) == TRUE )
```
Comment (vadz) [inline src/msw/printdlg.cpp:433]: This check doesn't make sense any more now:
```suggestion
        if ( printer->Open( printerName ) == TRUE )
```

File: src/msw/printdlg.cpp [hunk 3]
```diff
@@ -468,7 +474,7 @@ void wxWindowsPrintNativeData::InitializeDevMode(const wxString& printerName, Wi
         }
     }
 
-    if ( !m_devMode )
+    if ( !m_devMode && printerName.IsEmpty())
```
Comment (vadz) [inline src/msw/printdlg.cpp:477]: Another minor style fix:
```suggestion
    if ( !m_devMode && printerName.IsEmpty() )
```

**Review by vadz — COMMENTED** [review]
Thanks, this looks logical to me but I don't quite understand in which circumstances did we run into a problem with the old code: could you please give an example of a situation in which things didn't work before (and how?) and work correctly now to help me understand this?

TIA!

**Comment by sz5000** [comment]
The problem was the following:

if we call InitializeDevMode like this` InitializeDevMode("TestPrinter")`, then the printer parameter is NULL and this would execute the code block starting with the line  `PRINTDLG pd;`.  But this code block is just for fetching devmode for the default printer and if the default printer is not he printer we want, then we would initialize `m_devMode` with data from a wrong printer.

If we have initialized with a wrong devmode and if we then use `wxWindowsPrintNativeData::TransferFrom`,  which internally does the following
```
DocumentProperties( nullptr,
                printer,
                szPrinterName,
                (LPDEVMODE)hDevMode, // Reuse our buffer for output.
                (LPDEVMODE)hDevMode, // Pass the driver our changes
                DM_IN_BUFFER |       // Commands to Merge our changes and
                DM_OUT_BUFFER );     // write the result.
```

we pass a wrong devmode to the printer and this leads to a crash if the devmode is not like expected.

**Comment by sz5000** [comment]
If we have to initialize the devmode for a specific printer and if the wxPrinter argument is NULL which is the default, then the old code falls back to

```
PRINTDLG pd;

memset(&pd, 0, sizeof(PRINTDLG));
pd.lStructSize    = sizeof(PRINTDLG);

pd.hwndOwner      = nullptr;
pd.hDevMode       = nullptr; // Will be created by PrintDlg
pd.hDevNames      = nullptr; // Ditto

pd.Flags          = PD_RETURNDEFAULT;
```

This code just returns the devmode for the default printer because `PD_RETURNDEFAULT` is used here, but the default printer can also be a different printer than the one for which we need to initialize the devmode. Then we have a wrong devmode and apps can crash, just like in my APP.

**Comment by vadz** [comment]
Yes, I understood your explanation, thanks. What I was asking if it's really enough to select any non-default printer in the dialog to crash? Or if, perhaps, the hack in the commit I linked prevents the crash from happening?

I'd like to know this in order to write an accurate description of this change in the change log.

**Comment by sz5000** [comment]
It depends on the printer driver and the expected driver extra data. My app did not crash with the printers I have used so far, but recently I had to install OneNote because of an issue with the oneno

**Pull request closed without merging**